torchaudio/csrc/ffmpeg/buffer.cpp

```diff
@@ -0,0 +1,163 @@
+    }
+    case AV_SAMPLE_FMT_S32:
+    case AV_SAMPLE_FMT_S32P: {
+      t = torch::empty(shape, torch::kInt32);
+      ptr = reinterpret_cast<uint8_t*>(t.data_ptr<int32_t>());
+      break;
+    }
+    case AV_SAMPLE_FMT_S64:
+    case AV_SAMPLE_FMT_S64P: {
+      t = torch::empty(shape, torch::kInt64);
+      ptr = reinterpret_cast<uint8_t*>(t.data_ptr<int64_t>());
+      break;
+    }
+    case AV_SAMPLE_FMT_FLT:
+    case AV_SAMPLE_FMT_FLTP: {
+      t = torch::empty(shape, torch::kFloat32);
+      ptr = reinterpret_cast<uint8_t*>(t.data_ptr<float>());
+      break;
+    }
+    case AV_SAMPLE_FMT_DBL:
+    case AV_SAMPLE_FMT_DBLP: {
+      t = torch::empty(shape, torch::kFloat64);
+      ptr = reinterpret_cast<uint8_t*>(t.data_ptr<double>());
+      break;
+    }
+    default:
+      throw std::runtime_error(
+          "Unsupported audio format: " +
+          std::string(av_get_sample_fmt_name(format)));
+  }
+  for (int i = 0; i < num_planes; ++i) {
+    memcpy(ptr, pFrame->extended_data[i], plane_size);
+    ptr += plane_size;
+  }
+  if (is_planar)
+    t = t.t();
+  return t;
+}
+} // namespace
+
+void Buffer::push_audio_frame(AVFrame* pFrame) {
+  chunks.push_back(convert_audio_tensor(pFrame));
+}
+
+namespace {
+torch::Tensor convert_image_tensor(AVFrame* pFrame) {
+  // ref:
+  // https://ffmpeg.org/doxygen/4.1/filtering__video_8c_source.html#l00179
+  // https://ffmpeg.org/doxygen/4.1/decode__video_8c_source.html#l00038
+  AVPixelFormat format = static_cast<AVPixelFormat>(pFrame->format);
+  int width = pFrame->width;
+  int height = pFrame->height;
+  uint8_t* buf = pFrame->data[0];
+  int linesize = pFrame->linesize[0];
+
+  int channel;
+  switch (format) {
+    case AV_PIX_FMT_RGB24:
+    case AV_PIX_FMT_BGR24:
+      channel = 3;
+      break;
+    case AV_PIX_FMT_ARGB:
+    case AV_PIX_FMT_RGBA:
+    case AV_PIX_FMT_ABGR:
+    case AV_PIX_FMT_BGRA:
+      channel = 4;
+      break;
+    case AV_PIX_FMT_GRAY8:
+      channel = 1;
+      break;
+    default:
+      throw std::runtime_error(
+          "Unexpected format: " + std::string(av_get_pix_fmt_name(format)));
+  }
+
+  torch::Tensor t;
+  t = torch::empty({1, height, width, channel}, torch::kUInt8);
+  auto ptr = t.data_ptr<uint8_t>();
+  int stride = width * channel;
+  for (int i = 0; i < height; ++i) {
+    memcpy(ptr, buf, stride);
+    buf += linesize;
+    ptr += stride;
+  }
+  return t.permute({0, 3, 1, 2});
+}
+} // namespace
+
+void Buffer::push_video_frame(AVFrame* pFrame) {
+  chunks.push_back(convert_image_tensor(pFrame));
+}
+
+torch::Tensor Buffer::pop_all() {
+  if (!chunks.size())
+    return torch::empty({});
+
+  std::vector<torch::Tensor> tmp;
+  while (chunks.size()) {
+    tmp.push_back(chunks.front());
+    chunks.pop_front();
+  }
+  return torch::cat(tmp, 0);
+}
+
+void Buffer::push_frame(AVFrame* frame) {
+  switch (media_type) {
+    case AVMEDIA_TYPE_AUDIO:
+      push_audio_frame(frame);
+      break;
+    case AVMEDIA_TYPE_VIDEO:
```

Comment:
Where are ``AVMEDIA_TYPE_AUDIO`` and ``AVMEDIA_TYPE_VIDEO`` defined? Or it's taken care  by ffmpeg?

Reply:
https://ffmpeg.org/doxygen/4.1/group__lavu__misc.html#ga9a84bba4713dfced21a1a56163be1f48